### PR TITLE
Add a key binding Toggle Acrylic 

### LINKED ISSUE #2531: Add a key binding to set or toggle acrylic
Ability to quickly toggle opacity on/off through keystrokes. It would switch between 100% and custom setting in properties.

### PR DESCRIPTION
**Pull request by Jaswir**
## Summary of the Pull Request
Added a key binding Toggle Acrylic 

## References and Relevant Issues
#2531

## Detailed Description of the Pull Request / Additional comments
toggle acrylic v1

## Validation Steps Performed
Tested whether things are still working along with:
* Mouse scroll wheel
* Adjust opacity: image
* Acrylic Off when they turn off transparency.

In both situations with 
image

```
"useAcrylic": true 
"useAcrylic": false
```

## PR Checklist
- [ V] Closes #2531 
- [X ] Tests added/passed
- [ X] Documentation updated
   - If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/terminal) and link it here: #xxx
- [X] Schema updated (if necessary)

### CODE REVIEW

**Comment by Jaswir** [comment]
@zadjii-msft 

Why do I get the formatting error? I thought it was that I needed to remove the `\`
`    ON_ALL_ACTIONS(RestartConnection)       \`

File: src/cascadia/TerminalSettingsModel/AllShortcutActions.h [hunk 1]
```diff
@@ -106,7 +107,7 @@
     ON_ALL_ACTIONS(ShowContextMenu)         \
     ON_ALL_ACTIONS(ExpandSelectionToWord)   \
     ON_ALL_ACTIONS(CloseOtherPanes)         \
-    ON_ALL_ACTIONS(RestartConnection)
+    ON_ALL_ACTIONS(RestartConnection)
```
Comment (zadjii-msft) [inline src/cascadia/TerminalSettingsModel/AllShortcutActions.h:110]: ah, I bet it's complaining about the trailing spaces here. 


```
tools\runformat.cmd
```

should automatically reformat everything for you 😉

Comment (Jaswir) [reply src/cascadia/TerminalSettingsModel/AllShortcutActions.h:110]: image

Comment (Jaswir) [reply src/cascadia/TerminalSettingsModel/AllShortcutActions.h]: > ah, I bet it's complaining about the trailing spaces here.
> 
> ```
> tools\runformat.cmd
> ```
> 
> should automatically reformat everything for you 😉

Things are never easy are they. 
How can I solve. Noob at ps1 scripts but It seems like it should work cause the Invoke-CodeFormat function is inside the OpenConsole.psm1 script. As for now I went the manual route. Problem seems to be in ControlCore.cpp

Comment (DHowett) [reply src/cascadia/TerminalSettingsModel/AllShortcutActions.h]: Sorry about that.

Can you do this instead? It lets you do the steps from "runformat" manually.

It will be much easier than trying all your individual files to see which one was the problem.

```
import-module ./tools/openconsole.psm1
invoke-codeformat -ignorexaml
```

it will make changes to your code which you will need to commit. :smile:

Comment (Jaswir) [reply src/cascadia/TerminalSettingsModel/AllShortcutActions.h]: @zadjii-msft @DHowett 

I solved the issues. It's ready to be reviewed now.

**Comment by Jaswir** [comment]
Solved the issues. It's ready to be reviewed. Let me know if any changes need to be made.

**Comment by Jaswir** [comment]
@zadjii-msft @DHowett-MSFT @DHowett 

Hi Dustin, Hi Mike, 

I was wondering if you got a chance yet to review the code? 

@zadjii-msft 
Or would it stop being able to talk with you in a more comfortable manner about the other issue inside 2531, because you wouldn't get notifications anymore once it's merged or something?

File: src/cascadia/TerminalControl/ControlCore.h [hunk 2]
```diff
@@ -272,6 +274,8 @@ namespace winrt::Microsoft::Terminal::Control::implementation
             std::shared_ptr<ThrottledFuncTrailing<Control::ScrollPositionChangedArgs>> updateScrollBar;
         };
 
+        bool _acrylicToggle;
```
Comment (zadjii-msft) [inline src/cascadia/TerminalControl/ControlCore.h:277]: Why add another bool here, rather than leverage the existing `UseAcrylic` `RUNTIME_SETTING`?

**Review by zadjii-msft — COMMENTED** [review]
Yea I suppose I'm mostly just curious about the runtime setting thing. That seems like it would get us a lot of this for free.

Sorry for the delays! Been a busy summer 😅

File: src/cascadia/TerminalSettingsModel/defaults.json [hunk 3]
```diff
@@ -487,6 +487,21 @@
...
+                { "command": { "action": "adjustOpacity", "opacity": 0, "relative": false } },
+                { "command": { "action": "adjustOpacity", "opacity": 25, "relative": false } },
+                { "command": { "action": "adjustOpacity", "opacity": 50, "relative": false } },
+                { "command": { "action": "adjustOpacity", "opacity": 75, "relative": false } },
+                { "command": { "action": "adjustOpacity", "opacity": 100, "relative": false } }
+            ]
+
+            
+        },   
+        { "command": "toggleAcrylic", "keys": "ctrl+space" },
```
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/cascadia/TerminalSettingsModel/defaults.json:503]: actually I'm gonna block over this one

**Review by zadjii-msft — CHANGES_REQUESTED** [review]

**Comment by Jaswir** [comment]
> Yea I suppose I'm mostly just curious about the runtime setting thing. That seems like it would get us a lot of this for free.
> 
> Sorry for the delays! Been a busy summer 😅

@zadjii-msft 
I am worried that when this issue gets merged you won't get notification anymore for things said in 2531 chat. And I can't use it anymore to talk in a comfortable manner about the other issue. And feel like your response doesn't adress my concern.

So I would like to know if you still get notifications when something is said inside 2531 once it's merged.

**Comment by zadjii-msft** [comment]
> So I would like to know if you still get notifications when something is said inside https://github.com/microsoft/terminal/issues/2531 once it's merged

I get emails for every comment and on every issue, PR and discussion, and I try best to keep on top of them, no worries ☺️

File: src/cascadia/TerminalSettingsModel/Resources/en-US/Resources.resw [hunk 4]
```diff
@@ -571,6 +571,9 @@
     <value>Set background opacity to {0}%</value>
     <comment>A command to change how transparent the background of the window is</comment>
   </data>
+  <data name="ToggleAcrylicCommandKey" xml:space="preserve">
+    <value>Toggle Acrylic</value>
```
[the remark this answers is not in the document]
Comment (Jaswir) [reply src/cascadia/TerminalSettingsModel/Resources/en-US/Resources.resw:575]: Okay, how did I accept suggestions again..

[the remark this answers is not in the document]
Comment (Jaswir) [reply src/cascadia/TerminalSettingsModel/Resources/en-US/Resources.resw:575]: Ah I didn't know translators use those. Interesting fact

File: src/cascadia/TerminalControl/ControlCore.h (hunk 2 above)
Comment (Jaswir) [reply src/cascadia/TerminalControl/ControlCore.h:277]: Initially I used the `UseAcrylic RUNTIME_SETTING `and it seemed to work, but after testing Mouse Scroll Wheel to change Opacity and Adjust Opacity to change opacity, I stumbled upon some problems with this method.

**Problems:**
runtime_acrylic_breaks_adjustopac
runtime_acrylic_breaks_scrollwheel

**The actual Terminal Preview behaviour with UseAcrylic on in settings.json in these scenarios:**
runtime_acrylic_correct_adjustopacity
runtime_acrylic_correct_scrollwheel

When the user uses Mouse Scroll Wheel to change Opacity and Adjust Opacity to change opacity through the command palette the `ControlCore::_setOpacity `function is triggered. This function manually turns off the` _runtimeUseAcrylic` when opacity becomes 100. 

When the user scroll to opacity 100 and then scroll down again, the runtime acrylic goes false at 100 and would stay false when you go back under 100. Also when adjusting the opacity through the command palette similar thing happens when going to set background opacity to 100% option it stays false when going to 25%, 50%, etc.

Basically we can’t use the` UseAcrylic RUNTIME_SETTING`, because it becomes False when the user turns off transparency(opacity becomes 100)  by using Mouse Scroll Wheel to change Opacity and Adjust Opacity to change opacity and we can’t use `UseAcrylic RUNTIME_SETTING` to turn `_runtimeUseAcrylic `back on since it’s false. So I figured we need to use an external boolean besides the `UseAcrylic () RUNTIME_SETTING`.

Using an external boolean solves these issues.

Comment (Jaswir) [reply src/cascadia/TerminalControl/ControlCore.h:277]: How would you have solved this problem?

File: src/cascadia/TerminalSettingsModel/defaults.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (Jaswir) [reply src/cascadia/TerminalSettingsModel/defaults.json:503]: @zadjii-msft @kalimal

@zadjii-msft @KalleOlaviNiemitalo
> actually I'm gonna block over this one

I'm sorry not sure what you mean with that phrase to block over something. I understand that it means you need to think deeply about this, is that correct? 

I think it should have a key bind for sure because:

1. I am a Terminal user and want to be able to do this with a keybind. 
2. @moonunit9 would likely agree, given the initial ask, mentioning keystrokes: "Ability to quickly toggle opacity on/off through keystrokes. It would switch between 100% and custom setting in properties."
3. It is an Acrylic Toggle so you want to be able to switch it quickly a keybind allows for this. 
4. Try it yourself and see what you prefer, I am pretty sure you actually want it to be a keybind too. But ctrl+space might not be the best choice, maybe something else would be better? 

Could we maybe ask other terminal users somewhere about their opinion on this? When I made games I'd just ask the players for stuff like this, people's choice.

[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/cascadia/TerminalSettingsModel/defaults.json:503]: > > actually I'm gonna block over this one
> 
> I'm sorry not sure what you mean with that phrase: "to block over something".

Basically, in code-review parlance, "blocking" a PR is when a reviewer hits that "Request changes" button in response to an issue with the PR. By actually requesting changes, that prevents the PR from merging, even if two other people do sign off on it. 

----

It's totally fine for users to add a keybinding for this action, but I don't think we should bind one _for them_. They're always free to add their own keybinding in the settings. 

If you put it in `defaults.json` like this, then it'll bind <kbd>ctrl+space</kbd> to this action _for everyone_. Binding it here would prevent any CLI application from reading that key unless the user manually unbinds it. And <kbd>ctrl+space</kbd> is already used by different shells for a variety of things. Best we not steal it from them. 

If you change this instead to just

```json
{ "command": "toggleAcrylic" },
```

It'll still show up in the Command Palette, and in the Settings UI, so users can easily add their own keys.

File: src/cascadia/TerminalControl/ControlCore.h (hunk 2 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalControl/ControlCore.h:277]: Hmm, so I tried something in ffc78620e.

Instead of setting `_runtimeUseAcrylic` based on the opacity, when the opacity changes, just do the `opacity < 1.0` when we actually need it. 

That _seems_ to work... I'll play with it more but that's a start

File: src/cascadia/TerminalSettingsModel/defaults.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (Jaswir) [reply src/cascadia/TerminalSettingsModel/defaults.json:503]: Didn't know I could add new actions for things that had no keys bound to them. 
That wasn't intuitive to me. 

I see it now: 
image

[the remark this answers is not in the document]
Comment (Jaswir) [reply src/cascadia/TerminalSettingsModel/defaults.json]: @zadjii-msft 

Updated it

File: src/cascadia/TerminalControl/ControlCore.h (hunk 2 above)
Comment (Jaswir) [reply src/cascadia/TerminalControl/ControlCore.h:290]: @zadjii-msft Jaswir took 🗡️

Comment (Jaswir) [reply src/cascadia/TerminalControl/ControlCore.h:290]: @zadjii-msft Are you a DnD player?

**Review by lhecker — DISMISSED** [review]

File: src/cascadia/TerminalControl/ControlCore.h (hunk 2 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalControl/ControlCore.h:290]: I am, but that was more of a reference to 
image

Comment (Jaswir) [reply src/cascadia/TerminalControl/ControlCore.h:290]: I see, I didn't play much Zelda to be honest ..

**Comment by DHowett** [comment]
@zadjii-msft what happened with this one?
